adaptdl/adaptdl/torch/adascale.py

```diff
@@ -271,5 +329,12 @@ def patch_optimizer(self):
         @functools.wraps(self._optimizer.step)
         def wrapper(*args, **kwargs):
             return self.step(*args, **kwargs)
+        old_zero_grad = self._optimizer.zero_grad
 
+        @functools.wraps(self._optimizer.zero_grad)
+        def zero_wrapper(*args, **kwargs):
+            if self._made_step:
+                return old_zero_grad(*args, **kwargs)
+            return None
         self._optimizer.step = wrapper
+        self._optimizer.zero_grad = zero_wrapper
```

Comment:
I think there's also `model.zero_grad()` which might be used by some training code.

Reply:
Patched model.zero_grad() in `adaptdl/adaptdl/torch/parallel.py`